src/coreclr/jit/lower.cpp

```diff
@@ -7844,6 +7860,126 @@ void Lowering::ContainCheckBitCast(GenTree* node)
     }
 }
 
+//------------------------------------------------------------------------
+// LowerBlockStoreAsHelperCall: Lower a block store node as a memset/memcpy call
+//
+// Arguments:
+//    blkNode - The block store node to lower
+//
+void Lowering::LowerBlockStoreAsHelperCall(GenTreeBlk* blkNode)
+{
+    // We shouldn't be using helper calls for blocks on heap containing GC pointers.
+    // due to atomicity guarantees.
+    assert(!blkNode->IsZeroingGcPointersOnHeap());
+
+    LIR::Use use;
+    assert(!BlockRange().TryGetUse(blkNode, &use));
+
+    const bool isVolatile = blkNode->IsVolatile();
+
+    GenTree* dest = blkNode->Addr();
+    GenTree* data = blkNode->Data();
+    GenTree* size;
+
+    // Is it Memset ...
+    if (blkNode->OperIsInitBlkOp())
+    {
+        // Drop GT_INIT_VAL nodes
+        if (data->OperIsInitVal())
+        {
+            BlockRange().Remove(data);
+            data = data->gtGetOp1();
+        }
+    }
+    else
+    {
+        // ... or Memcpy?
+        assert(data->OperIs(GT_IND, GT_LCL_VAR, GT_LCL_FLD));
+
+        // Drop GT_IND nodes
+        if (data->OperIs(GT_IND))
+        {
+            BlockRange().Remove(data);
+            data = data->AsIndir()->Addr();
+        }
+
+        if (varTypeIsStruct(data))
+        {
+            // If it's a struct local, take its address
+            const unsigned lclNum = data->AsLclVarCommon()->GetLclNum();
+            GenTreeLclFld* addr   = comp->gtNewLclAddrNode(lclNum, data->AsLclVarCommon()->GetLclOffs(), TYP_BYREF);
+            BlockRange().InsertAfter(data, addr);
+            BlockRange().Remove(data);
+            data = addr;
+        }
+    }
+
+    if (blkNode->OperIs(GT_STORE_DYN_BLK))
+    {
+        // Size is not a constant
+        size = blkNode->AsStoreDynBlk()->gtDynamicSize;
+    }
+    else
+    {
+        // Size is a constant
+        size = comp->gtNewIconNode(blkNode->Size(), TYP_I_IMPL);
+        BlockRange().InsertBefore(data, size);
+    }
+
+    // A hacky way to safely call fgMorphTree in Lower
+    GenTree* destPlaceholder = comp->gtNewZeroConNode(dest->TypeGet());
+    GenTree* dataPlaceholder = comp->gtNewZeroConNode(data->TypeGet());
+    GenTree* sizePlaceholder = comp->gtNewZeroConNode(size->TypeGet());
+
+    CorInfoHelpFunc helper = blkNode->OperIsInitBlkOp() ? CORINFO_HELP_MEMSET : CORINFO_HELP_MEMCPY;
+    GenTreeCall* call = comp->gtNewHelperCallNode(helper, TYP_VOID, destPlaceholder, dataPlaceholder, sizePlaceholder);
+    comp->fgMorphTree(call);
+
+    LIR::Range range      = LIR::SeqTree(comp, call);
+    GenTree*   rangeStart = range.FirstNode();
+    GenTree*   rangeEnd   = range.LastNode();
+
+    BlockRange().InsertBefore(blkNode, std::move(range));
+    blkNode->gtBashToNOP();
```

Comment:
```suggestion
    blkNode->gtBashToNOP();
```

Just remove it? We don't generally create nops in lowering.

Reply:
this will require a bit of work since the caller will then use gtNext to lower next node. I tried to propagate gtNext as a return result but then I realized that StoreBlk can be lowered from multiple places and all of them don't expect it to disappear.